### PR TITLE
NNI Engine: Compute NNI Likelihoods

### LINKED ISSUE #383: Be able to access the NNI count from Python
I'd like to be able to get the count of NNIs from the DAG from within Python.

### LINKED ISSUE #405: NNI Engine: Compute Likelihoods for NNIs.
Dependent on issue #404.  Once we have a resizeable GPEngine, we need to use it to compute the likelihoods for all adjacent NNIs, to determine which to accept and which to reject.

### PR DESCRIPTION
**Pull request by davidrich27**
## Description

Compute NNI Likelihoods via NNIEngine
-  Iterates over the Accepted NNI's in NNI Engine and builds GPOperationVector which computes all likelihoods in series.
-  For each NNI Likelihood, uses knowledge of NNI's DAG neighbor's PLVs pre-existing in the DAG in order to compute NNI likelihood with minimal PLV updating.

Miscellaneous
- Makes Adjacent NNI Count accessible through pybito.
- Adds function for fully connecting SubsplitDAG -- all legal edges connecting any two nodes which are not already present are added to the DAG.

Closes #405 
Closes #383 

## Tests

- TEST_CASE("NNI Engine: NNI Likelihoods")
  - Compares resulting NNI Likelihoods of two methods: 
    - via inserting NNI into DAG, updating all PLVs, and recomputing likelihoods (used as truth).
    - via NNIEngine. 

## Checklist:

* [x] Code follows our detailed [contribution guidelines](CONTRIBUTING.md)
* [x] `clang-format` has been run
* [x] TODOs have been eliminated from the code
* [x] Comments are up to date, document intent, and there are no commented-out code blocks

### CODE REVIEW

File: src/nni_engine.hpp [hunk 1]
```diff
@@ -60,48 +68,141 @@ class NNIEngine {
...
-  double ComputeNNILikelihood(const NNIOperation &nni);
-  // Add score to given NNI.
-  void AddScoreForNNI(const NNIOperation &nni, const double score);
-  double GetScoreForNNI(const NNIOperation &nni) const;
+  // ** Filter Functions
 
-  // ** Filtering
+  // Initialization step for filter before beginning
+  using StaticFilterInitFunction =
+      std::function<void(NNIEngine &, GPEngine &, GraftDAG &)>;
```
Comment (matsen) [inline src/nni_engine.hpp:75]: This seems great to me, though I want to double check with @ognian- who has some opinions regarding passing functions. We can discuss tomorrow.

File: src/nni_engine.hpp [hunk 2]
```diff
@@ -60,48 +68,141 @@ class NNIEngine {
...
+  static const size_t KeyIndexCount = 9;
+  using KeyIndexIterator =
+      EnumIterator<KeyIndex, KeyIndex::Parent_Id, KeyIndex::Child_PHatRight>;
+  using KeyIndexMap = EnumArray<KeyIndex, KeyIndexCount, size_t>;
+  using KeyIndexMapPair = std::pair<KeyIndexMap, KeyIndexMap>;
+  using KeyIndexPairArray = std::array<std::pair<KeyIndex, KeyIndex>, 3>;
+
+  // Translate NNI clade type to key index type.
+  using NNIClade = NNIOperation::NNIClade;
+  static KeyIndex NNICladeToKeyIndex(NNIClade clade_type) {
```
Comment (matsen) [inline src/nni_engine.hpp:123]: This function is specific to PHat

File: src/nni_engine.hpp [hunk 3]
```diff
@@ -132,30 +231,112 @@ class NNIEngine {
...
-  void ClearRejectedNNIs();
+
+  // Add score to given NNI.
+  void AddScoreForNNI(const NNIOperation &nni, const double score);
+  // Get score for given NNI.
+  double GetScoreForNNI(const NNIOperation &nni) const;
+
+  // Remove all adjacent nnis from list. Option to keep and re-evaluate NNIs from
+  // previous pass.
+  void UpdateAdjacentNNIs(const bool reevaluate_rejected_nnis = false);
```
Comment (matsen) [inline src/nni_engine.hpp:247]: Perhaps rename to something with Remove?

File: src/nni_operation.hpp [hunk 4]
```diff
@@ -19,15 +19,29 @@
...
 class NNIOperation {
  public:
-  NNIOperation() : parent_(0), child_(0){};
-  NNIOperation(Bitset parent, Bitset child) : parent_(parent), child_(child){};
+  NNIOperation() : parent_(0), child_(0), is_focal_clade_on_right_(){};
+
+  NNIOperation(Bitset parent, Bitset child) : parent_(parent), child_(child) {
+    const SubsplitClade clade =
+        Bitset::SubsplitIsChildOfWhichParentClade(parent_, child_);
+    const bool is_focal_on_right = (clade == SubsplitClade::Right);
```
Comment (matsen) [inline src/nni_operation.hpp:37]: compactify

**Review by matsen — COMMENTED** [review]
Use Ognian's clade definition?

File: src/eigen_sugar.hpp [hunk 5]
```diff
@@ -58,37 +58,53 @@ inline EigenVectorXd EigenVectorXdOfStdVectorDouble(std::vector<double> &v) {
...
     }
+    CHECK_LT(error, tolerance);
   }
-  return true;
 };
 
-void CheckVectorXdEquality(const EigenVectorXd v1, const EigenVectorXd v2,
-                           double tolerance) {
-  CHECK_EQ(v1.size(), v2.size());
+// Return the maximum error between any two cells.
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/eigen_sugar.hpp]: Updated comments.

File: src/gp_engine.hpp [hunk 6]
```diff
@@ -60,16 +72,26 @@ class GPEngine {
...
   void SetTransitionAndDerivativeMatricesToHaveBranchLength(double branch_length);
   void SetTransitionMatrixToHaveBranchLengthAndTranspose(double branch_length);
-  const Eigen::Matrix4d& GetTransitionMatrix() { return transition_matrix_; };
   void SetBranchLengths(EigenVectorXd branch_lengths);
   void SetBranchLengthsToConstant(double branch_length);
+  void SetBranchLengthsToDefault();
   void ResetLogMarginalLikelihood();
-  double GetLogMarginalLikelihood() const;
 
+  void CopyNodeData(const size_t src_node_idx, const size_t dest_node_idx);
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/gp_engine.hpp:87]: Added comments.

File: src/gp_engine.hpp [hunk 7]
```diff
@@ -91,15 +117,23 @@ class GPEngine {
   EigenConstVectorXdRef GetHybridMarginals() const;
   EigenConstVectorXdRef GetSBNParameters() const;
 
+  double GetLogMarginalLikelihood() const;
+  const Eigen::Matrix4d& GetTransitionMatrix() { return transition_matrix_; };
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/gp_engine.hpp]: Made const.

File: src/nni_engine.hpp (hunk 2 above)
Comment (davidrich27) [reply src/nni_engine.hpp]: Renamed function to `NNICladeToPHatPLV`

File: src/nni_operation.hpp (hunk 4 above)
Comment (davidrich27) [reply src/nni_operation.hpp]: Done.

File: src/nni_engine.hpp (hunk 3 above)
Comment (davidrich27) [reply src/nni_engine.hpp:242]: Added this clarifying comment to what the function is doing.  I think that "Update" is appropriate, but tell me what you think.
  
  // Update adjacent NNIs at end of current sweep.  If not re-evaluating rejected NNIs,
  // then current adjacent NNIs are removed.  Then NNIs that are adjacent to last sweep's
  // accepted NNIs are added.

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,5 +1,12 @@
 cmake_minimum_required(VERSION 3.19)
 
+# disallow building in same directory as source.
```
[the remark this answers is not in the document]
Comment (ognian-) [reply CMakeLists.txt:3]: Looks fine to me too.

**Review by matsen — APPROVED** [review]
Great work! Just one little FYI here.

File: src/gp_doctest.cpp [hunk 9]
```diff
@@ -1244,117 +1247,119 @@ TEST_CASE("GPEngine: Resize and Reindex GPEngine after AddNodePair") {
...
+    test_array.push_back(test_passes);
+    // Finally, test run full GP Optimization after all modifications completed.
+    inst.ReinitializePriors();
+    dag.ReinitializeTidyVectors();
+    inst.EstimateBranchLengths(0.0001, 100, true);
+    inst.PopulatePLVs();
+    inst.ComputeLikelihoods();
+    inst.ComputeMarginalLikelihood();
+
+    test_passes = true;
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/gp_doctest.cpp]: Good change!

**Pull request merged**